pybaseball/datasources/fangraphs.py

```diff
@@ -1,44 +1,135 @@
+              league: FanGraphsLeague = FanGraphsLeague.ALL, qual: Optional[int] = None, split_seasons: bool = True,
+              month: FanGraphsMonth = FanGraphsMonth.ALL, on_active_roster: bool = False, minimum_age: int = MIN_AGE,
+              maximum_age: int = MAX_AGE, team: str = '0', _filter: str = '', players: str = '',
+              position: FanGraphsPositions = FanGraphsPositions.ALL, max_results: int = 1000000,) -> pd.DataFrame:
 
-    return fg_data
+        types = self.DEFAULT_TYPES if types is None else types
 
+        if start_season is None:
+            raise ValueError(
+                "You need to provide at least one season to collect data for. " +
+                "Try specifying start_season or start_season and end_season."
+            )
 
-def get_fangraphs_tabular_data_from_url(url: str, column_name_mapper: Callable = None, known_percentages: List[str] = []) -> pd.DataFrame:
-    content = requests.get(ROOT_URL + url).content
+        if end_season is None:
+            end_season = start_season
 
-    return get_fangraphs_tabular_data_from_html(
-        content,
-        column_name_mapper=column_name_mapper,
-        known_percentages=known_percentages
-    )
+        assert self.STATS is not None
+
+        url_options = {
+            'pos': position.value,
+            'stats': self.STATS.value,
+            'lg': league.value,
+            'qual': qual if qual is not None else 'y',
+            'type': types,
+            'season': end_season,
+            'month': month.value,
+            'season1': start_season,
+            'ind': int(split_seasons),
+            'team': team + ',ts' if self.TEAM_DATA else '',
+            'rost': int(on_active_roster),
+            'age': f"{minimum_age},{maximum_age}",
+            'filter': _filter,
+            'players': players,
+            'page': f'1_{max_results}'
+        }
+
+        return self._validate(
+            self._postprocess( 
+                self.html_accessor.get_tabular_data_from_options(
+                    self.QUERY_ENDPOINT,
+                    query_params=url_options,
+                    # TODO: Remove the type: ignore after this is fixed: https://github.com/python/mypy/issues/5485
+                    column_name_mapper=self.COLUMN_NAME_MAPPER, # type: ignore
+                    known_percentages=self.KNOWN_PERCENTAGES,
+                )
+            )
+        )
+
+class FanGraphsBattingDataTable(FanGraphsDataTable):
+    STATS: FanGraphsStatTypes = FanGraphsStatTypes.BATTING
+    DEFAULT_TYPES: FanGraphsStat = FanGraphsBattingStat.ALL()
+    COLUMN_NAME_MAPPER: ColumnListMapperFunction = BattingStatsColumnMapper().map_list
+    KNOWN_PERCENTAGES: List[str] = ["GB/FB"]
+
+    def _postprocess(self, data: pd.DataFrame) -> pd.DataFrame:
+        return data.sort_values(["WAR", "OPS"], ascending=False)
+
+class FanGraphsPitchingDataTable(FanGraphsDataTable):
+    STATS: FanGraphsStatTypes = FanGraphsStatTypes.PITCHING
```

Comment:
I'm finding it a little confusing that `STATS` has a data type `FanGraphsStatTypes` and `TYPES` has a data type `FanGraphsStat`. Could we somehow make the names more consistent here? One thought is maybe the argument names for the method don't have to match the names in the fangraphs API? i.e. what the `types` query parameter is specifying is maybe better thought of table columns than "stats" (or "types" which is super ambiguous not tp mention conflict with the python protected keyword `type`)?

Reply:
Valid point. I'll rework.

Reply:
Names have been reworked to be less confusing.